### ISSUE TITLE
unsorted CGUs

### ISSUE THREAD

**Issue by nnethercote** [rust-lang/rust#113425]
I slightly changed how CGUs are named in #112946 and the fuchsia developers reported an ICE on their Windows CI builds. I created #113402 to give a more informative error message.

@ComputerDruid responded with an updated crash [here](https://github.com/rust-lang/rust/pull/112946#issuecomment-1624237973). I'm now moving this problem into its own issue.

**Comment by nnethercote** [comment]
[Responding to [this comment](https://github.com/rust-lang/rust/pull/112946#issuecomment-1624237973) from the original PR]

> And when I tried switching `jemalloc=false` on my linux machine, I don't encounter the problem.

When you have `jemalloc=true` on linux do you encounter the problem?

> ```
> error: internal compiler error: compiler\rustc_monomorphize\src\partitioning.rs:195:9: unsorted CGUs:
>                                 - pkg_config.1b83623108ca023c-cgu.0
>                                 - pkg_config.1b83623108ca023c-cgu.1
>                                 - pkg_config.1b83623108ca023c-cgu.2
>                                 - pkg_config.1b83623108ca023c-cgu.3
>                                 - pkg_config.1b83623108ca023c-cgu.4
>                                 - pkg_config.1b83623108ca023c-cgu.5
>                                 - pkg_config.1b83623108ca023c-cgu.6
>                                 - pkg_config.1b83623108ca023c-cgu.7
>                                 - pkg_config.1b83623108ca023c-cgu.8
>                                 - pkg_config.1b83623108ca023c-cgu.9
>                                 - pkg_config.1b83623108ca023c-cgu.10
> 
> error: internal compiler error: compiler\rustc_monomorphize\src\partitioning.rs:195:9: unsorted CGUs:
>                                 - vcpkg.da648edadde02677-cgu.0
>                                 - vcpkg.da648edadde02677-cgu.1
>                                 - vcpkg.da648edadde02677-cgu.2
>                                 - vcpkg.da648edadde02677-cgu.3
>                                 - vcpkg.da648edadde02677-cgu.4
>                                 - vcpkg.da648edadde02677-cgu.5
>                                 - vcpkg.da648edadde02677-cgu.6
>                                 - vcpkg.da648edadde02677-cgu.7
>                                 - vcpkg.da648edadde02677-cgu.8
>                                 - vcpkg.da648edadde02677-cgu.9
>                                 - vcpkg.da648edadde02677-cgu.10
>                                 - vcpkg.da648edadde02677-cgu.11
>                                 - vcpkg.da648edadde02677-cgu.12
>                                 - vcpkg.da648edadde02677-cgu.13
>                                 - vcpkg.da648edadde02677-cgu.14
>                                 - vcpkg.da648edadde02677-cgu.15
> ```

Thanks, this is exactly the information I was after. Something weird is going on here -- those numeric suffixes are supposed to be zero-padded, e.g. `00` to `10` for the first examples, and `00` to `15` for the second example. Because then they'd be in name-sorted order. (Without the zero-padding, the one with the `10` suffix is considered lower than the one with the `9` suffix.)

The code responsible for creating the CGU name is [here](https://github.com/rust-lang/rust/blob/85bf07972a1041b9e25393b803d0e006bec3eaaf/compiler/rustc_monomorphize/src/partitioning.rs#L418-L447). It uses `let num_digits = codegen_units.len().ilog10() as usize + 1;` to compute how many digits the suffix must be, and then creates the suffix with `let suffix = format!("{index:0num_digits$}");`. Nothing too complicated, except for the use of the magic trailing `$` to allow a variable padding amount. And platform independent. The number of CGUs involved in the examples (11 and 16) are completely normal, anything up to 16 is normal, though the code should handle any number. Seems like the kind of code that should not be failing on a single project's CI on a single platform!

Makes me wonder if there could be a bug on Windows with either `ilog10` or the use of `0num_digits$` padding in `format!`?

**Comment by ComputerDruid** [comment]
> When you have jemalloc=true on linux do you encounter the problem?

No, I just tried that too and neither way can I reproduce the issue on my linux machine

**Comment by lqd** [comment]
@ComputerDruid your [config](https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8776317651163748305/+/u/generate_config.toml/raw_io.output_text?format=raw) uses ThinLTO and LLD: I believe you're hitting https://github.com/rust-lang/rust/issues/109114.

Removing `lto = "thin"` will fix it.
